### PR TITLE
[TensorIR][M2a] Compute-Inline,Reverse-Compute-Inline

### PR DESCRIPTION
**Pull request by junrushao1994**
This PR is part of the TensorIR upstreaming effort (#7527), which adds the first 2 schedule primitives:
- compute-Inline
- reverse-compute-inline

### CODE REVIEW

File: src/tir/schedule/utils.h [hunk 1]
```diff
@@ -114,6 +116,33 @@ inline bool CanRelaxStorageUndereThread(const runtime::StorageScope& storage_sco
...
+/******** SeqStmt ********/
+
+/*!
+ * \brief Remove a specific Stmt from a SeqStmt. If a SeqStmt contains a BlockRealize,
+ * whose block is the Stmt to be removed, then remove that BlockRealize too.
+ * \param seq The SeqStmt to be removed from
+ * \param to_remove The Stmt to be removed
+ * \return The removal result
+ */
+inline Stmt SeqStmtRemove(const SeqStmt& seq, const Stmt& to_remove) {
```
Comment (comaniac) [inline src/tir/schedule/utils.h:128]: Ditto. `RemoveSeqStmt`?

**Review by comaniac — COMMENTED** [review]
LGTM. Mostly nits.

File: src/support/array.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+/*!
+ * \brief Checks if two arrays contain the same objects
+ * \tparam T The type of objects in the array
+ * \param a The first array
+ * \param b The second array
+ * \return A boolean indicating if they are the same
+ */
+template <class T>
+inline bool ArraySame(const Array<T>& a, const Array<T>& b) {
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply src/support/array.h:36]: Thanks for the suggestion! I will go with `IsSameArray`

File: src/tir/schedule/utils.h (hunk 1 above)
Comment (junrushao1994) [reply src/tir/schedule/utils.h:128]: What about `RemoveFromSeqStmt`?

File: src/tir/schedule/analysis.h [hunk 3]
```diff
@@ -41,8 +41,51 @@ void VerifySRefTree(const ScheduleState& self);
...
+/******** Scope ********/
+/*!
+ * \brief Get the sref to the scope root block, exclusive
+ * \param sref The block or loop sref to be retrieved
+ * \return The sref to the scope root block. NullOpt if `sref` is the root block of the IR
+ */
+Optional<StmtSRef> GetScopeRoot(const StmtSRef& sref);
+
+/*!
+ * \brief Checks if scope the specified sref is in is a stage-pipeline and return it
```
[the remark this answers is not in the document]
Comment (junrushao1994) [reply src/tir/schedule/analysis.h:53]: @tkonolige once told me that we should use "Checks" (IIRC), but I agree we should append "if so" to the end of the sentence

[the remark this answers is not in the document]
Comment (comaniac) [reply src/tir/schedule/analysis.h:53]: Actually both are fine to me as long as it is consistent. For example, the above `GetScopeRoot` has "Get the sref...". It should be "Gets the sref" then.

File: src/tir/schedule/utils.h (hunk 1 above)
Comment (comaniac) [reply src/tir/schedule/utils.h:128]: Sounds good.

File: src/tir/schedule/analysis.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (junrushao1994) [reply src/tir/schedule/analysis.h:53]: Got it. I am going to change all of them to "Gets"

File: src/support/array.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jcf94) [reply src/support/array.h:36]: How about to merge this as a member function of `Array`? Some object of TVM has member functions like `A.same_as(B)`.

[the remark this answers is not in the document]
Comment (tqchen) [reply src/support/array.h:36]: same_as is a shallow ptr check which is different. ArrayWithSameContents might be a good way to differentiate

[the remark this answers is not in the document]
Comment (junrushao1994) [reply src/support/array.h:36]: Sounds good. I will go with ArrayWithSameContent

**Comment by junrushao1994** [comment]
Hey sorry for the late response! I addressed the comments accordingly. Please take another look :-)

**Review by comaniac — APPROVED** [review]
LGTM

**Review by jcf94 — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**